diff --git a/src/core/timeouts.c b/src/core/timeouts.c
index f37acfec..71db3c7d 100644
--- a/src/core/timeouts.c
+++ b/src/core/timeouts.c
@@ -74,7 +74,7 @@
 
 /** This array contains all stack-internal cyclic timers. To get the number of
  * timers, use LWIP_ARRAYSIZE() */
-const struct lwip_cyclic_timer lwip_cyclic_timers[] = {
+const struct lwip_cyclic_timer lwip_cyclic_timers[] __attribute__((section( "\".irom.text.lwip2\""))) = {
 #if LWIP_TCP
   /* The TCP timer is a special case: it does not have to run always and
      is triggered to start from TCP using tcp_timer_needed() */
